### ISSUE TITLE
Curse of Recklessness doesn't stop fleeing target. It works to p...

### ISSUE THREAD

**Issue by kaytotes** [Project-Epoch/bugtracker#7695]
> Curse of Recklessness doesn't stop fleeing target. It works to prevent flee if the target is already debuffed when it tries to flee, but doesnt do anything if the flee/fear is already on. This is wrong, Reck must cancel any flee effect asap

.go xyz -484.428000 -1424.250000 89.480000 0 0.588631

**Comment by zana244** [comment]
Confirmed on internal. Same behaviour should be for judgement of justice.
Should also interact with warlock fear -- allowing warlock to ping pong target.
#2544

**Comment by kaytotes** [comment]
Retest

**Comment by Shadownflame** [comment]
Properly interacts with warlock fear now (the ping pong mechanic).
Bug persists with mob AI flee effect still. If you pre-apply the COR, the mob won't start fleeing when they fall low health, but if they start fleeing, applying COR won't stop them.

Judgment of Justice behaves the same way.

**Comment by T1ti** [comment]
@Shadownflame can you confirm if this comment in the code is true ? https://github.com/Project-Epoch/TrinityCore/pull/31/files#diff-38ecc72a17e3d339e5d66ae517c1c94a6497e84101046fed7f287bc6ec91f470R2894

Does it only happen if they can find an ally ?
If so I assume it's because their state changes from fleeing to like running to ally

**Comment by Shadownflame** [comment]
No, it doesn't stop the mob flee at all. If there is no ally around at all, the mob just runs around aimlessly, but still continues fleeing. It doesn't even kick in when they randomly change directions every 3-4sec.

**Comment by T1ti** [comment]
https://github.com/T1ti/TrinityCore/commits/titi_3.3.5
Got a potential fix, works for me

Here's the issue detailed if it can help figuring out a better fix

- Regular fears set unit state UNIT_STATE_FLEEING which applies a regular **not timed** fleeing motion(removed when spell expires) `FLEEING_MOTION_TYPE`

- Fleeing for assistance `(Creature::DoFleeToGetAssistance())` sets can set two other different states based on if a creature is in range or not : 

1. If a creature is in range it does this : 
MotionMaster::MoveSeekAssistance(creaturepos) which sets `UNIT_STATE_ROAMING_MOVE`, starts a `ASSISTANCE_MOTION_TYPE ` which basically tells it to run to the creature and do a call for help when reaching point.
 
2. if not it does : SetControlled(true, UNIT_STATE_FLEEING); which is pretty simple and similar to regular fleeing, with the difference that it doesn't have a base timer because it's not from a spell/aura with a duration so it calls  `MotionMaster::MoveFleeing(Unit* enemy, uint32 time)` with this time a timed fleeingMotion called `TIMED_FLEEING_MOTION_TYPE` and 5000msec (config) and applies unit state UNIT_STATE_FLEEING_MOVE 

So yes, there are **three** different motions generators AND unit states for those three cases.
**UNIT_STATE_FLEEING  + FLEEING_MOTION_TYPE
UNIT_STATE_ROAMING_MOVE + ASSISTANCE_MOTION_TYPE
UNIT_STATE_FLEEING_MOVE + TIMED_FLEEING_MOTION_TYPE**

An issue I had with ASSISTANCE_MOTION_TYPE if that when forceably cancelled early, unit would get stuck and do nothing, I sovled that issue by commenting the `movementInform `check in `AssistanceMovementGenerator::Finalize` , **also it's up to you if you want creatures to still do a call for help at the position they get cancelled or not**  _(anyways pretty sure creatures only doing a call for help when reachnig point isn't blizzlike, isn't it periodic?)_

Note : currently how it works is that the spell will stop any unit using UNIT_STATE_ROAMING_MOVE + ASSISTANCE_MOTION_TYPE , if this is ever used by some scripts or whatever warlock and paladins can potentially cancel those.
We could potentially also apply the state `UNIT_STATE_FLEEING_MOVE `when calling DoFleeToGetAssistance too to check for this specific case, but I don't think assist is really used for anythnig else than that

**Comment by zana244** [comment]
What's more, on internal, if you remove CoR or SoJ by casting a different rank or a different seal/curse the mob starts to flee, but it can still attack you if it randomly walks near you, without a fear effect on them :P

**Comment by T1ti** [comment]
I assume this doesn't happen with base fear type and only with low health flee ?

**Comment by zana244** [comment]
regarding my comment only, it doesn't even need a fear effect on the mob ^_^.
If you randomly cast CoR/SoJ on a 100% hp mob for example, and then you remove the effect, it just starts fleeing (it looks like just the fleeing movement generator is being used) but the mob can still attack you if it paths next to you.
